File: sdk/go/examples/webhook_history.go
// Webhook history example — fetch recent webhook deliveries and inspect payload variants.
//
// Usage:
//
//	PACHCA_TOKEN=your_token go run webhook_history.go
package main

import (
	"context"
	"fmt"
	"log"
	"os"

	pachca "github.com/pachca/openapi/sdk/go/generated"
)

func main() {
	token := os.Getenv("PACHCA_TOKEN")
	if token == "" {
		log.Fatal("Set PACHCA_TOKEN environment variable")
	}

	client := pachca.NewPachcaClient(token)
	limit := int32(5)
	response, err := client.Bots.GetWebhookEvents(context.Background(), &pachca.GetWebhookEventsParams{
		Limit: &limit,
	})
	if err != nil {
		log.Fatal(err)
	}

	fmt.Printf("Fetched %d webhook events\n", len(response.Data))
	for i, event := range response.Data {
		fmt.Printf("%d. id=%s created_at=%s payload=%s\n", i+1, event.ID, event.CreatedAt.Format("2006-01-02T15:04:05Z07:00"), summarizePayload(event.Payload))
	}

	fmt.Printf("has_next=%v next_page=%q\n", boolValue(response.Meta.Paginate.HasNext), response.Meta.Paginate.NextPage)
}

func boolValue(value *bool) bool {
	return value != nil && *value
}

func summarizePayload(payload pachca.WebhookPayloadUnion) string {
	switch {
	case payload.LinkSharedWebhookPayload != nil:
		p := payload.LinkSharedWebhookPayload
		return fmt.Sprintf("link_shared message_id=%d links=%d user_id=%d", p.MessageID, len(p.Links), p.UserID)
	case payload.MessageWebhookPayload != nil:
		p := payload.MessageWebhookPayload
		return fmt.Sprintf("message event=%s id=%d chat_id=%d", p.Event, p.ID, p.ChatID)
	case payload.ReactionWebhookPayload != nil:
		p := payload.ReactionWebhookPayload
		return fmt.Sprintf("reaction event=%s message_id=%d code=%s", p.Event, p.MessageID, p.Code)
	case payload.ButtonWebhookPayload != nil:
		p := payload.ButtonWebhookPayload
		return fmt.Sprintf("button message_id=%d user_id=%d", p.MessageID, p.UserID)
	case payload.ViewSubmitWebhookPayload != nil:
		p := payload.ViewSubmitWebhookPayload
		return fmt.Sprintf("view user_id=%d fields=%d", p.UserID, len(p.Data))
	case payload.ChatMemberWebhookPayload != nil:
		p := payload.ChatMemberWebhookPayload
		return fmt.Sprintf("chat_member event=%s chat_id=%d users=%d", p.Event, p.ChatID, len(p.UserIDs))
	case payload.CompanyMemberWebhookPayload != nil:
		p := payload.CompanyMemberWebhookPayload
		return fmt.Sprintf("company_member event=%s users=%d", p.Event, len(p.UserIDs))
	default:
		return "unknown"
	}
}

File: sdk/kotlin/examples/webhook_history.kt
/**
 * Webhook history example — fetch recent webhook deliveries and inspect payload variants.
 *
 * Usage:
 *
 *   PACHCA_TOKEN=your_token kotlin webhook_history.kt
 */
package examples.webhookhistory

import com.pachca.sdk.*
import kotlinx.coroutines.runBlocking

fun main() = runBlocking {
    val token = System.getenv("PACHCA_TOKEN")
        ?: error("Set PACHCA_TOKEN environment variable")

    val client = PachcaClient(token)
    val response = client.bots.getWebhookEvents(limit = 5)

    println("Fetched ${response.data.size} webhook events")
    response.data.forEachIndexed { index, event ->
        println("${index + 1}. id=${event.id} created_at=${event.createdAt} payload=${summarizePayload(event.payload)}")
    }

    println("has_next=${response.meta.paginate.hasNext} next_page=${response.meta.paginate.nextPage}")
    client.close()
}

private fun summarizePayload(payload: WebhookPayloadUnion): String = when (payload) {
    is LinkSharedWebhookPayload -> "link_shared message_id=${payload.messageId} links=${payload.links.size} user_id=${payload.userId}"
    is MessageWebhookPayload -> "message event=${payload.event} id=${payload.id} chat_id=${payload.chatId}"
    is ReactionWebhookPayload -> "reaction event=${payload.event} message_id=${payload.messageId} code=${payload.code}"
    is ButtonWebhookPayload -> "button message_id=${payload.messageId} user_id=${payload.userId}"
    is ViewSubmitWebhookPayload -> "view user_id=${payload.userId} fields=${payload.data.size}"
    is ChatMemberWebhookPayload -> "chat_member event=${payload.event} chat_id=${payload.chatId} users=${payload.userIds.size}"
    is CompanyMemberWebhookPayload -> "company_member event=${payload.event} users=${payload.userIds.size}"
}
